New testing framework for the incremental compiler

Duhemm opened the pull request:
The tests that are included in this PR are juste examples. I'll start porting some of the scripted tests if this new format LGTY!

Duhemm committed:
New testing framework for the incremental compiler

The goal of this testing framework is to give a mean to concisely write
complete scenarios for the incremental compiler, like scripted tests,
but solely focused on the incremental compiler.

jsuereth commented on internal/compile-inc/src/test/scala/sbt/inc/IncrementalCompilerTest.scala:
```diff
@@ -0,0 +1,254 @@
+import xsbt.{ CompilationFailedException, TestAnalyzingCompiler }
+
+object IncrementalCompilerTest {
+  implicit class FileOP(val content: String)
```
So we lift strings into FileOPs?

Duhemm replied:
Yes, this allows us to write something like `"A.scala" -> delete` to express file deletion.

jsuereth commented:
SO, to help be maven frinedly, I think we need to remove the `test->test` dependnecy.

Instead, create a new "incrementalCOmpilerTestFramework" project which has your test-harness code and then depend on that with `test->compile`. 

Otherwise, I love the new test DSL, looks great!

Duhemm commented:
@jsuereth Would it be enough to just change the dependency marked `test->test` to `test->compile`?

Duhemm committed:
Fix dependency between projects

Duhemm committed:
Migrate tests for incremental compiler to Scalatest

This commit also fixes some bugs in `TestAnalysisCompiler` and
`TestAnalysisCallback` so that their behavior correctly mimics the
behavior of the incremental compiler.

Duhemm commented:
@jsuereth @eed3si9n Any news on this PR?

jsuereth commented:
yeah, this LGTM assuming we enable TravisCI.

Duhemm committed:
Add support for pending scenarios

Duhemm committed:
Fixes to incremental compiler testing

Duhemm committed:
Convenience conversion to mark tests as pending

Duhemm committed:
Show compilation errors on unexpected compilation failure

Duhemm committed:
Keep last changes after failed compilation

Duhemm committed:
Port scripted tests relative to incremental compiler

Duhemm committed:
Fix bug with incremental compiler tests with Scala 2.11

If a file was changed, failed to compile and changed immediately again,
then the changes to the file were considered twice. We should only keep
the latest modification.

Duhemm committed:
Move sources for compiler-bridge, fix test dependencies

The new tests for the incremental compiler use some classes defined in
the compiler bridge. Because there were two different projects for the
compiler bridge, that means that we would have had to depend on one of
the two projects based on the value of `scalaVersion`. This kind of
dependency is not allowed by sbt.

The solution is to merge the two projects and specify the location of
the sources within this project based on the value of  `scalaVersion`.

Duhemm committed:
Mark 2 tests as pending

These 2 tests have mistakenly been marked as passing because they
seemed to have been fixed in Scala 2.11, but were not.

Duhemm commented:
I've rebased this PR and changed a few things regarding the splitting of the compiler interface. These new tests use classes defined in the compiler bridge. Because there were 2 projects, we would have needed to depend on either one based on the value of `scalaVersion`. Instead I merged the two projects in one, and made `scalaSource` depend on the value of  `scalaVersion`.
